viewing_party/party.py

```diff
@@ -1,23 +1,184 @@
 # ------------- WAVE 1 --------------------
 
 def create_movie(title, genre, rating):
-    pass
+    movie_dict = {}
+
+    if not title or not genre or not rating:
+        return None
+    else:
+        movie_dict['title'] = title
+        movie_dict['genre'] = genre
+        movie_dict['rating'] = rating
+    
+    return movie_dict
+
+def add_to_watched(user_data, movie):
+
+    user_data["watched"].append(movie)
+    
+    return user_data
+
+def add_to_watchlist(user_data, movie):
+
+    user_data["watchlist"].append(movie)
+    
+    return user_data
+
+def watch_movie(user_data, title):
+
+    for movie in user_data["watchlist"]:
+        if movie["title"] == title:
+            user_data["watchlist"].remove(movie)
+            user_data["watched"].append(movie)
+            break
+    return user_data
 
 # -----------------------------------------
 # ------------- WAVE 2 --------------------
 # -----------------------------------------
 
+def get_watched_avg_rating(user_data):
+
+    if not user_data["watched"]:
+        return 0.0
+    
+    total_rating = 0
+    for movie in user_data["watched"]:
+        total_rating += movie["rating"] 
+            
+    average = total_rating / len(user_data["watched"])
+
+    return average
+
+def get_most_watched_genre(user_data):
+    genre_count = {"Fantasy": 0, "Intrigue": 0, "Action": 0} #change method so that other genres can be included
+
+    if not user_data["watched"]:
+        return None 
+    
+    for movie in user_data["watched"]:
+        genre = movie["genre"]
+        if genre in genre_count:
+            genre_count[genre] += 1   
+    top_genre = max(genre_count, key=genre_count.get)
+
+    return top_genre
 
 # -----------------------------------------
 # ------------- WAVE 3 --------------------
 # -----------------------------------------
+def get_unique_watched(user_data):
+
+    movies_not_watched_by_friends = []
+
+    for movie_watched_by_user in user_data["watched"]:
+        watched_by_any_friend = False
+
+        # checks if the movie is has been watched by a friend
+        for friend in user_data["friends"]:
+            if movie_watched_by_user in friend["watched"]:
+                watched_by_any_friend = True
+                break
+
+        # If no friend has watched it, adds it to the final list
+        if not watched_by_any_friend:
+            movies_not_watched_by_friends.append(movie_watched_by_user)
+    
+    return movies_not_watched_by_friends

```

Comment:
Great job on completing this function! If I could offer a piece a feedback it would be maybe flattening the lists of `friends` and their lists of watched movies. That way you can just loop over the a master collection of friends' movies instead of looping over the friends and then their list of movies. We could even make the master collection a `dict` so don't have to loop over that. The implementation could look like this:
```python
def get_unique_watched(user_data):
    user_watched = user_data["watched"]
    friends = user_data["friends"]
    friend_titles = {}
    unique_movies = []

    for friend in friends:
        watched_movies = friend["watched"]
        
        for movie in watched_movies:
            title = movie["title"]
            friend_titles[title] = True
    
    for movie in user_watched:
        title = movie["title"]

        if not friend_titles.get(title):
            unique_movies.append(movie)
    
    return unique_movies
```

Reply:
What do you think the time complexity of implementation like this would be?